File: frontend/src/api/shopping.ts
import type {
  ShoppingCreateFromPlanPayload,
  ShoppingItemPatchPayload,
  ShoppingListBulkDeletePayload,
  ShoppingListBulkDeleteResponse,
  ShoppingListMergePayload,
  ShoppingListRead,
  ShoppingListSummary,
  ShoppingManualItemCreatePayload,
} from "../types/shopping";
import { ApiError, apiRequest } from "./http";

const TOKEN_KEY = "access_token";

function getToken() {
  return localStorage.getItem(TOKEN_KEY);
}

async function requestWithApiError<T>(request: Promise<T>): Promise<T> {
  try {
    return await request;
  } catch (err) {
    if (err instanceof ApiError) throw err;
    throw new ApiError(0, err instanceof Error ? err.message : "Не удалось выполнить запрос.", err);
  }
}

export async function listShoppingLists(): Promise<ShoppingListSummary[]> {
  return requestWithApiError(
    apiRequest<ShoppingListSummary[]>({
      method: "GET",
      path: "/shopping-lists",
      token: getToken(),
    }),
  );
}

export async function createShoppingListFromPlan(payload: ShoppingCreateFromPlanPayload): Promise<ShoppingListRead> {
  return requestWithApiError(
    apiRequest<ShoppingListRead>({
      method: "POST",
      path: "/shopping-lists/from-plan",
      token: getToken(),
      body: payload,
    }),
  );
}

export async function getShoppingList(shoppingListId: number | string): Promise<ShoppingListRead> {
  return requestWithApiError(
    apiRequest<ShoppingListRead>({
      method: "GET",
      path: `/shopping-lists/${shoppingListId}`,
      token: getToken(),
    }),
  );
}

export async function deleteShoppingList(shoppingListId: number | string): Promise<void> {
  await requestWithApiError(
    apiRequest<void>({
      method: "DELETE",
      path: `/shopping-lists/${shoppingListId}`,
      token: getToken(),
    }),
  );
}

export async function bulkDeleteShoppingLists(
  payload: ShoppingListBulkDeletePayload,
): Promise<ShoppingListBulkDeleteResponse> {
  return requestWithApiError(
    apiRequest<ShoppingListBulkDeleteResponse>({
      method: "POST",
      path: "/shopping-lists/bulk-delete",
      token: getToken(),
      body: payload,
    }),
  );
}

export async function mergeShoppingLists(payload: ShoppingListMergePayload): Promise<ShoppingListRead> {
  return requestWithApiError(
    apiRequest<ShoppingListRead>({
      method: "POST",
      path: "/shopping-lists/merge",
      token: getToken(),
      body: payload,
    }),
  );
}

export async function rebuildShoppingList(shoppingListId: number | string): Promise<ShoppingListRead> {
  return requestWithApiError(
    apiRequest<ShoppingListRead>({
      method: "POST",
      path: `/shopping-lists/${shoppingListId}/rebuild`,
      token: getToken(),
    }),
  );
}

export async function patchShoppingItem(
  shoppingListId: number | string,
  itemId: number | string,
  payload: ShoppingItemPatchPayload,
): Promise<void> {
  await requestWithApiError(
    apiRequest<void>({
      method: "PATCH",
      path: `/shopping-lists/${shoppingListId}/items/${itemId}`,
      token: getToken(),
      body: payload,
    }),
  );
}

export async function createManualShoppingItem(
  shoppingListId: number | string,
  payload: ShoppingManualItemCreatePayload,
): Promise<void> {
  await requestWithApiError(
    apiRequest<void>({
      method: "POST",
      path: `/shopping-lists/${shoppingListId}/items/manual`,
      token: getToken(),
      body: payload,
    }),
  );
}

export async function deleteShoppingListItem(
  shoppingListId: number | string,
  itemId: number | string,
): Promise<void> {
  await requestWithApiError(
    apiRequest<void>({
      method: "DELETE",
      path: `/shopping-lists/${shoppingListId}/items/${itemId}`,
      token: getToken(),
    }),
  );
}

File: backend/app/api/routes/shopping.py
from fastapi import APIRouter, Depends, HTTPException, Response, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.models.user import User
from app.schemas.shopping import (
    ShoppingListBulkDeleteRequest,
    ShoppingListBulkDeleteResponse,
    ShoppingListCreateFromPlanRequest,
    ShoppingListItemRead,
    ShoppingListItemUpdate,
    ShoppingListMergeRequest,
    ShoppingListRead,
    ShoppingListSummaryRead,
    ShoppingManualItemCreate,
)
from app.services.shopping import (
    ShoppingItemUpdateForbiddenError,
    ShoppingListItemNotFoundError,
    ShoppingListNotFoundError,
    ShoppingListSourceNotSupportedError,
    ShoppingPlanNotFoundError,
    add_manual_item,
    create_shopping_list_from_plan,
    delete_shopping_list,
    delete_shopping_lists,
    delete_shopping_list_item,
    get_shopping_list,
    list_shopping_lists,
    merge_shopping_lists,
    rebuild_shopping_list_from_sources,
    update_shopping_list_item,
)

router = APIRouter(prefix="/shopping-lists", tags=["shopping-lists"])


@router.get("", response_model=list[ShoppingListSummaryRead])
def get_shopping_lists(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return list_shopping_lists(db, current_user.id)


@router.post("/from-plan", response_model=ShoppingListRead, status_code=status.HTTP_201_CREATED)
def post_shopping_list_from_plan(
    payload: ShoppingListCreateFromPlanRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return create_shopping_list_from_plan(db, current_user.id, payload)
    except ShoppingPlanNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Plan not found") from exc


@router.post("/merge", response_model=ShoppingListRead, status_code=status.HTTP_201_CREATED)
def post_merge_shopping_lists(
    payload: ShoppingListMergeRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return merge_shopping_lists(db, current_user.id, payload)
    except ShoppingListNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Shopping list not found") from exc
    except ShoppingPlanNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Plan not found") from exc


@router.post("/bulk-delete", response_model=ShoppingListBulkDeleteResponse)
def post_bulk_delete_shopping_lists(
    payload: ShoppingListBulkDeleteRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return delete_shopping_lists(db, current_user.id, payload)
    except ShoppingListNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Shopping list not found") from exc


@router.get("/{shopping_list_id}", response_model=ShoppingListRead)
def get_shopping_list_by_id(
    shopping_list_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return get_shopping_list(db, current_user.id, shopping_list_id)
    except ShoppingListNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Shopping list not found") from exc


@router.delete("/{shopping_list_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_shopping_list_by_id(
    shopping_list_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        delete_shopping_list(db, current_user.id, shopping_list_id)
    except ShoppingListNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Shopping list not found") from exc
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{shopping_list_id}/rebuild", response_model=ShoppingListRead)
def post_rebuild_shopping_list(
    shopping_list_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return rebuild_shopping_list_from_sources(db, current_user.id, shopping_list_id)
    except ShoppingListNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Shopping list not found") from exc
    except ShoppingPlanNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Plan not found") from exc
    except ShoppingListSourceNotSupportedError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc)) from exc


@router.patch("/{shopping_list_id}/items/{item_id}", response_model=ShoppingListItemRead)
def patch_shopping_list_item(
    shopping_list_id: int,
    item_id: int,
    payload: ShoppingListItemUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return update_shopping_list_item(db, current_user.id, shopping_list_id, item_id, payload)
    except ShoppingListNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Shopping list not found") from exc
    except ShoppingListItemNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Shopping list item not found") from exc
    except ShoppingItemUpdateForbiddenError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc)) from exc


@router.post("/{shopping_list_id}/items/manual", response_model=ShoppingListItemRead, status_code=status.HTTP_201_CREATED)
def post_manual_shopping_item(
    shopping_list_id: int,
    payload: ShoppingManualItemCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return add_manual_item(db, current_user.id, shopping_list_id, payload)
    except ShoppingListNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Shopping list not found") from exc


@router.delete("/{shopping_list_id}/items/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_shopping_item(
    shopping_list_id: int,
    item_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        delete_shopping_list_item(db, current_user.id, shopping_list_id, item_id)
    except ShoppingListNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Shopping list not found") from exc
    except ShoppingListItemNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Shopping list item not found") from exc
    return Response(status_code=status.HTTP_204_NO_CONTENT)
